test/mpi/threads/perf/mt_rma_msgrate.c

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+/* This program provides a simple test of send-receive performance between
+   two (or more) processes.  This sometimes called head-to-head or
+   ping-ping test, as both processes send at the same time.
+*/
+
+#include "mpi.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <assert.h>
+#include "mpitest.h"
+#include "mpithreadtest.h"
+
+#define CACHELINE_SIZE 64
+
+#define MESSAGE_SIZE 8
+#define NUM_MESSAGES 64000
+#define WINDOW_SIZE 64
+
+#define ERROR_MARGIN 0.05       /* FIXME: a better margin? */
+
+enum RMA_OPs {
+    OP_PUT,
+    OP_GET,
+    OP_ACC,
+    OP_GACC,
+    OP_INVALID
+};
+
+int rma_op = OP_INVALID;
+int world_rank;
+MPI_Comm *thread_wins;
+double *t_elapsed;
+
+MTEST_THREAD_RETURN_TYPE thread_fn(void *arg);
+
+MTEST_THREAD_RETURN_TYPE thread_fn(void *arg)
+{
+    int error;
+    int tid;
+    MPI_Win my_win;
+    int win_i, win_post_i, win_posts;
+    void *buf, *result_buf;
+    MPI_Request requests[WINDOW_SIZE];
+    MPI_Status statuses[WINDOW_SIZE];
+    double t_start, t_end;
+
+    tid = (int) (long) arg;
+    my_win = thread_wins[tid];
+
+    win_posts = NUM_MESSAGES / WINDOW_SIZE;
+    assert(win_posts * WINDOW_SIZE == NUM_MESSAGES);
+
+    /* Allocate a cache-aligned buffer to prevent potential effects of serialization:
+     * either false-sharing on the CPU or serialization in the NIC's parallel TLB
+     * engine
+     */
+    error = posix_memalign(&buf, CACHELINE_SIZE, MESSAGE_SIZE * sizeof(char));
```

Comment:
Should this be page aligned rather that just cacheline?

Reply:
This is the origin buffer, which is local to the process, right? What is the concern that benefits from page alignment?

Reply:
I read your comment above about the "NIC's parallel TLB engine", which would work at page granularity. Though I'm not sure if there's actually any concern with this usage.

Reply:
I copied from Rohit's `mt_pt2pt_msgrate.c` -- without actually reading it :(. Yeah, I think it is worthwhile to be page-aligned.

Reply:
Updated both `mt_pt2pt_msgrate` and `mt_rma_msgrate` to use page-size alignment. Even if it doesn't matter, it is good to eliminate potential concerns in a test.